compiler/src/org.graalvm.compiler.lir.aarch64/src/org/graalvm/compiler/lir/aarch64/AArch64ArrayIndexOfOp.java

```diff
@@ -77,190 +90,301 @@ public AArch64ArrayIndexOfOp(int arrayBaseOffset, JavaKind valueKind, LIRGenerat
+        masm.neon.andVVV(ASIMDSize.FullReg, chunkPart2RegV, chunkPart2RegV, tmpRegV1);
+        /* 4.2 Convert 32-byte to 8-byte representation, 2 bits per byte. */
+        /* Reduce from 256 -> 128 bits. */
+        masm.neon.addpVVV(ASIMDSize.FullReg, ElementSize.Byte, chunkPart1RegV, chunkPart1RegV, chunkPart2RegV);
+        /*
+         * Reduce from 128 -> 64 bits. Note tmpRegV1's value doesn't matter; only care about
+         * chunkPart1RegV.
+         */
+        masm.neon.addpVVV(ASIMDSize.FullReg, ElementSize.Byte, chunkPart1RegV, chunkPart1RegV, tmpRegV1);
 
-        masm.bind(match);
-        if (isUTF16) {
-            // Convert byte offset of searchChar to its char index in UTF-16 string
-            masm.asr(64, curIndex, curIndex, 1);
+        try (ScratchRegister scratchReg = masm.getScratchRegister()) {
+            Register matchPositionReg = scratchReg.getRegister();
+            /* Detect the byte starting position of matching element in the chunk. */
+            masm.neon.moveFromIndex(ElementSize.DoubleWord, ElementSize.DoubleWord, matchPositionReg, chunkPart1RegV, 0);
+            masm.rbit(64, matchPositionReg, matchPositionReg);
+            masm.clz(64, matchPositionReg, matchPositionReg);
+            masm.add(64, result, currOffset, matchPositionReg, ShiftType.ASR, 1);
+        }
+        if (shiftSize != 0) {
+            /* Convert byte offset of searchElement to its array index */
+            masm.asr(64, result, result, shiftSize);
         }
-        masm.add(64, result, endIndex, curIndex);
+
         masm.bind(end);
     }
+
+    @Override
+    public void emitCode(CompilationResultBuilder crb, AArch64MacroAssembler masm) {
+        Register result = asRegister(resultValue);
+        Register arrayLength = asRegister(arrayLengthValue);
+        Register fromIndex = asRegister(fromIndexValue);
+        Register baseAddress = asRegister(temp1);
+        Register searchLength = asRegister(temp2);
+
+        Label searchByChunk = new Label();
+        Label ret = new Label();
+
+        /*
+         * @formatter:off
+         * The arguments satisfy the following constraints:
+         *  1. 0 <= fromIndex <= arrayLength - 1 (or 2 when findTwoConsecutive is true)
+         *  2. number of characters in searchChar is 1 (or 2 when findTwoConsecutive is true)
+         * @formatter:on
+        */
+        masm.mov(result, -1);   // Return for empty strings
+        masm.cbz(64, arrayLength, ret);
+
+        /* Load address of first array element */
+        masm.add(64, baseAddress, asRegister(arrayPtrValue), arrayBaseOffset);
+
+        /*
+         * Search element-by-element for small arrays (with search space size of less than 32 bytes,
+         * i.e., 4 UTF-16 or 8 Latin1 elements) else search chunk-by-chunk.
+         */
+        masm.sub(64, searchLength, arrayLength, fromIndex);
+        if (!findTwoConsecutive) {
+            int chunkByteSize = 32;
```

Comment:
You had this as 48. Is there a reason this isn't 32?

Reply:
Oh yes, 48 turned out to be a sweet spot to switch to chunk-based approach in micro-benchmarking on N1 (about 4.5% improvement relative to the time it took with size 32). Other systems that I tested, including A72, also showed  improvements (relatively small though, less than 2% relative to time for 32) for longer strings. However, there is no single threshold that I found suitable across AArch64 machines to switch from char-by-char to chunk-based search. The thresholds mostly varied between 48 and 64. 
Well, micro-benchmarking numbers may not be enough to rely upon. I used them as a hint that the higher threshold of 48 would compensate for additional complexity of SIMD instructions and its approach of detecting a match. I'm fine with the threshold set to 32 as well.

Reply:
Let's keep it as 32. The difference between 32 and 48 as thresholds is now less than the noise (~1.6% with 32 being faster) on the synthetic microbenchmark. 
The newer version is now 66% faster relative to master on N1.